crd/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use stackable_operator::Crd;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::str::FromStr;
+
+// TODO: We need to validate the name of the cluster because it is used in pod and configmap names, it can't bee too long
+// This probably also means we shouldn't use the node_names in the pod_name...
+
+#[derive(Clone, CustomResource, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[kube(
+    group = "spark.stackable.de",
+    version = "v1",
+    kind = "SparkCluster",
+    shortname = "sc",
+    namespaced
+)]
+#[kube(status = "SparkClusterStatus")]
+pub struct SparkClusterSpec {
+    pub master: SparkNode,
+    pub worker: SparkNode,
+    pub history_server: Option<SparkNode>,
+    pub image: String,
+    pub secret: Option<String>,
+    pub log_dir: Option<String>,
+}
+
+impl SparkClusterSpec {
+    /// collect hashed selectors from master, worker and history server
+    pub fn get_hashed_selectors(
+        &self,
+        cluster_name: &str,
+    ) -> HashMap<SparkNodeType, HashMap<String, SparkNodeSelector>> {
+        let mut hashed_selectors: HashMap<SparkNodeType, HashMap<String, SparkNodeSelector>> =
+            HashMap::new();
+
+        hashed_selectors.insert(
+            SparkNodeType::Master,
+            self.master
+                .get_hashed_selectors(SparkNodeType::Master, cluster_name),
+        );
+
+        hashed_selectors.insert(
+            SparkNodeType::Worker,
+            self.worker
+                .get_hashed_selectors(SparkNodeType::Worker, cluster_name),
+        );
+
+        if let Some(history_server) = &self.history_server {
+            hashed_selectors.insert(
+                SparkNodeType::HistoryServer,
+                history_server.get_hashed_selectors(SparkNodeType::HistoryServer, cluster_name),
+            );
+        }
+
+        hashed_selectors
+    }
+}
+
+#[derive(Clone, Debug, Hash, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+pub struct SparkNode {
+    pub selectors: Vec<SparkNodeSelector>,
+    // options
+    // master -> use Option<T>
+
+    // worker -> use Option<T>
+
+    // history_server -> use Option<T>
+}
+
+#[derive(Derivative, Clone, Debug, Deserialize, Eq, JsonSchema, Serialize)]
+#[derivative(Hash, PartialEq)]
+pub struct SparkNodeSelector {
+    // common
+    pub node_name: String,
+    // we need to ignore instances from hashing -> if we scale up or down, the hash
+    // for the selectors changes and the operator removes all nodes and rebuilds them
+    #[derivative(Hash = "ignore", PartialEq = "ignore")]
+    pub instances: usize,
+    pub config: Option<Vec<ConfigOption>>,
+    pub env: Option<Vec<ConfigOption>>,
+    // master -> use Option<T>
+
+    // worker -> use Option<T>
+    pub cores: Option<usize>,
+    pub memory: Option<String>,
+    // history_server -> use Option<T>
```

Comment:
leftover?

Reply:
Its a reminder where and how to add new configuration features (will make TODOs out of that)